crates/ty_python_semantic/src/types/narrow.rs

```diff
@@ -1416,6 +1431,17 @@ impl<'db, 'ast> NarrowingConstraintsBuilder<'db, 'ast> {
             ) {
                 constraints.insert(place, constraint);
             }
+
+            // Narrow tagged unions of tuples with `Literal` elements, just like `if` statements.
+            if let Some((place, constraint)) = self.narrow_tuple_subscript(
```

Comment:
we can probably skip this entirely if we already added a typeddict tagged-union constraint immediately above

```suggestion
            else if let Some((place, constraint)) = self.narrow_tuple_subscript(
```